### PR TITLE
Lora for SFT/SteerLM

### PR DESCRIPTION
**Pull request by HeyyyyyyG**
# What does this PR do ?

Add LoRA support for SFT/SteerLM
Only support DP and TP
Have to change Dockerfile to cherry pick a new commit in NeMo to support O2

# Changelog 
- Please update the [CHANGELOG.md](/CHANGELOG.md) under next version with high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation? Make sure to also update the [NeMo Framework User Guide](https://docs.nvidia.com/nemo-framework/user-guide/latest/index.html) which contains the tutorials

# Checklist when contributing a new algorithm
- [ ] Does the trainer resume and restore model state all states?
- [ ] Does the trainer support all parallelism techniques(PP, TP, DP)?
- [ ] Does the trainer support `max_steps=-1` and `validation`?
- [ ] Does the trainer only call APIs defined in [alignable_interface.py](/nemo_aligner/models/alignable_interface.py)?
- [ ] Does the trainer have proper logging?

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo_aligner/utils/utils.py
Comment (gshennvm) [inline nemo_aligner/utils/utils.py:100]: if you want the model_cfg can you use this function instead?

[load_checkpoint_model_config](https://github.com/NVIDIA/NeMo-Aligner/blob/main/nemo_aligner/utils/utils.py#L96)

File: nemo_aligner/utils/train_script_utils.py
Comment (gshennvm) [inline nemo_aligner/utils/train_script_utils.py:98]: nit: can this be moved to the bottom of the init_using_ptl function?

File: examples/nlp/gpt/train_gpt_sft.py
Comment (gshennvm) [inline examples/nlp/gpt/train_gpt_sft.py:155]: just making sure -- when PTL resumes it will automatically figure out how to load the LoRa weights and overwrite whatever we did here right?

Comment (gshennvm) [inline examples/nlp/gpt/train_gpt_sft.py:155]: i believe we also want to use PEFT for other algorithms as well, so I think we should create a `initialize_peft` function in `train_script_utils` to handle all this logic related to PEFT(such as error checking, and loading the adaptors)

File: examples/nlp/gpt/train_gpt_sft.py
Comment (gshennvm) [inline examples/nlp/gpt/train_gpt_sft.py:129]: maybe the `initialize_peft` function should take in a peft config

File: examples/nlp/gpt/train_gpt_sft.py
Comment (gshennvm) [inline examples/nlp/gpt/train_gpt_sft.py:133]: I think we shouldn't have to raise this on aligner in case nemo ever fixes it. @cuichenx will do a check for this on the nemo side. we can pick him his commit once done

**Review by gshennvm — CHANGES_REQUESTED** [review]
thanks for the change! i added some comments, and once we resolve we should add some small tests onto the CI

File: Dockerfile
Comment (gshennvm) [inline Dockerfile:59]: in addition, picking in this commit will be good https://github.com/NVIDIA/NeMo/pull/8323/files

File: examples/nlp/gpt/train_gpt_sft.py
Comment (HeyyyyyyG) [reply examples/nlp/gpt/train_gpt_sft.py]: Cool, didn't notice the check in nemo

Comment (HeyyyyyyG) [reply examples/nlp/gpt/train_gpt_sft.py]: 605bad8baf13e6f5107f9e330f79b269fb8866dd

File: examples/nlp/gpt/train_gpt_sft.py
Comment (HeyyyyyyG) [reply examples/nlp/gpt/train_gpt_sft.py]: 09ad8c61852debbbd2e713396f4b99f1441b52f5

File: examples/nlp/gpt/train_gpt_sft.py
Comment (HeyyyyyyG) [reply examples/nlp/gpt/train_gpt_sft.py]: Good point, 09ad8c61852debbbd2e713396f4b99f1441b52f5

Comment (HeyyyyyyG) [reply examples/nlp/gpt/train_gpt_sft.py]: yes

File: nemo_aligner/utils/train_script_utils.py
Comment (HeyyyyyyG) [reply nemo_aligner/utils/train_script_utils.py]: No we can't because it will cause an error when resume training. I move it to `init_peft` function, maybe that looks better? 
605bad8baf13e6f5107f9e330f79b269fb8866dd

File: nemo_aligner/utils/utils.py
Comment (HeyyyyyyG) [reply nemo_aligner/utils/utils.py:100]: What we need here is the model_cfg after applying `modify_config_fn` , I don't see how we can replace with this function

File: nemo_aligner/utils/train_script_utils.py
Comment (gshennvm) [inline nemo_aligner/utils/train_script_utils.py:140]: nit: can this updated_cfg be called `peft_cfg`?

File: examples/nlp/gpt/train_gpt_sft.py
Comment (gshennvm) [inline examples/nlp/gpt/train_gpt_sft.py:139]: we should just give the peft cfg

File: nemo_aligner/utils/utils.py
Comment (gshennvm) [reply nemo_aligner/utils/utils.py:100]: ah sorry good point -- the SFT configs work a bit differently than the other trainers(a long standing TODO to fix). this shouldn't block your PR so resolving it

File: examples/nlp/gpt/train_gpt_sft.py
Comment (HeyyyyyyG) [reply examples/nlp/gpt/train_gpt_sft.py:139]: Unfortunately this is not only peft cfg, it also contains base model cfg, because we need those (e.g num_attention_heads, hidden_size, num_layers) to initialize lora modules in [nemo](https://github.com/NVIDIA/NeMo/blob/5fdd12e9a9711b241023eb4ed0922733d69ded5e/nemo/collections/nlp/parts/peft_config.py#L69).

Comment (gshennvm) [reply examples/nlp/gpt/train_gpt_sft.py:139]: ah okay. then i'm good with this.

**Comment by gshennvm** [comment]
BEFORE MERGE: I am triggering a pipeline to check SFT with your commit. I'll approve once it passes.

**Review by gshennvm — APPROVED** [review]
tests passed thanks!

**Pull request merged**
